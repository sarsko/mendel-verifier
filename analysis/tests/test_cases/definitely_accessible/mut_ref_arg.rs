#[analyzer::run]
fn mut_ref_arg(x: &mut u32) {
    *x -= 1;
    *x += 1;
}

#[analyzer::run]
fn mut_ref_in_arg(x: (u32, (&mut u32, u32))) {
    *x.1.0 -= 1;
    *x.1.0 += 1;
}

fn main() {}
