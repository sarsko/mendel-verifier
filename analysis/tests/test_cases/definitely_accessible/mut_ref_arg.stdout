Analyzing file $DIR/mut_ref_arg.rs using DefinitelyAccessibleAnalysis...
Result for function mut_ref_arg():
{
  "bb0": [
    [
      [
        "state:",
        {
          "accessible": [
            "_1"
          ],
          "owned": [
            "_1"
          ]
        },
        "statement: _2 = CheckedSub((*_1), const 1_u32)"
      ]
    ],
    "state before terminator:",
    {
      "accessible": [
        "_1",
        "_2"
      ],
      "owned": [
        "_1",
        "_2"
      ]
    },
    "terminator: assert(!move (_2.1: bool), /"attempt to compute `{} - {}`, which would overflow/", (*_1), const 1_u32) -> [success: bb1, unwind: bb3]",
    {
      "bb1": [
        "state:",
        {
          "accessible": [
            "_1",
            "_2"
          ],
          "owned": [
            "_1",
            "_2"
          ]
        }
      ],
      "bb3": [
        "state:",
        {
          "accessible": [],
          "owned": []
        }
      ]
    }
  ],
  "bb1": [
    [
      [
        "state:",
        {
          "accessible": [
            "_1",
            "_2"
          ],
          "owned": [
            "_1",
            "_2"
          ]
        },
        "statement: (*_1) = move (_2.0: u32)"
      ],
      [
        "state:",
        {
          "accessible": [
            "_1",
            "_2"
          ],
          "owned": [
            "_1",
            "_2"
          ]
        },
        "statement: _3 = CheckedAdd((*_1), const 1_u32)"
      ]
    ],
    "state before terminator:",
    {
      "accessible": [
        "_1",
        "_2",
        "_3"
      ],
      "owned": [
        "_1",
        "_2",
        "_3"
      ]
    },
    "terminator: assert(!move (_3.1: bool), /"attempt to compute `{} + {}`, which would overflow/", (*_1), const 1_u32) -> [success: bb2, unwind: bb3]",
    {
      "bb2": [
        "state:",
        {
          "accessible": [
            "_1",
            "_2",
            "_3"
          ],
          "owned": [
            "_1",
            "_2",
            "_3"
          ]
        }
      ],
      "bb3": [
        "state:",
        {
          "accessible": [],
          "owned": []
        }
      ]
    }
  ],
  "bb2": [
    [
      [
        "state:",
        {
          "accessible": [
            "_1",
            "_2",
            "_3"
          ],
          "owned": [
            "_1",
            "_2",
            "_3"
          ]
        },
        "statement: (*_1) = move (_3.0: u32)"
      ],
      [
        "state:",
        {
          "accessible": [
            "_1",
            "_2",
            "_3"
          ],
          "owned": [
            "_1",
            "_2",
            "_3"
          ]
        },
        "statement: _0 = const ()"
      ]
    ],
    "state before terminator:",
    {
      "accessible": [
        "_0",
        "_1",
        "_2",
        "_3"
      ],
      "owned": [
        "_0",
        "_1",
        "_2",
        "_3"
      ]
    },
    "terminator: return",
    {}
  ],
  "bb3": [
    [],
    "state before terminator:",
    {
      "accessible": [],
      "owned": []
    },
    "terminator: resume",
    {}
  ]
}
Result for function mut_ref_in_arg():
{
  "bb0": [
    [
      [
        "state:",
        {
          "accessible": [
            "_1"
          ],
          "owned": [
            "_1"
          ]
        },
        "statement: _2 = CheckedSub((*((_1.1: (&mut u32, u32)).0: &mut u32)), const 1_u32)"
      ]
    ],
    "state before terminator:",
    {
      "accessible": [
        "_1",
        "_2"
      ],
      "owned": [
        "_1",
        "_2"
      ]
    },
    "terminator: assert(!move (_2.1: bool), /"attempt to compute `{} - {}`, which would overflow/", (*((_1.1: (&mut u32, u32)).0: &mut u32)), const 1_u32) -> [success: bb1, unwind: bb3]",
    {
      "bb1": [
        "state:",
        {
          "accessible": [
            "_1",
            "_2"
          ],
          "owned": [
            "_1",
            "_2"
          ]
        }
      ],
      "bb3": [
        "state:",
        {
          "accessible": [],
          "owned": []
        }
      ]
    }
  ],
  "bb1": [
    [
      [
        "state:",
        {
          "accessible": [
            "_1",
            "_2"
          ],
          "owned": [
            "_1",
            "_2"
          ]
        },
        "statement: (*((_1.1: (&mut u32, u32)).0: &mut u32)) = move (_2.0: u32)"
      ],
      [
        "state:",
        {
          "accessible": [
            "_1",
            "_2"
          ],
          "owned": [
            "_1",
            "_2"
          ]
        },
        "statement: _3 = CheckedAdd((*((_1.1: (&mut u32, u32)).0: &mut u32)), const 1_u32)"
      ]
    ],
    "state before terminator:",
    {
      "accessible": [
        "_1",
        "_2",
        "_3"
      ],
      "owned": [
        "_1",
        "_2",
        "_3"
      ]
    },
    "terminator: assert(!move (_3.1: bool), /"attempt to compute `{} + {}`, which would overflow/", (*((_1.1: (&mut u32, u32)).0: &mut u32)), const 1_u32) -> [success: bb2, unwind: bb3]",
    {
      "bb2": [
        "state:",
        {
          "accessible": [
            "_1",
            "_2",
            "_3"
          ],
          "owned": [
            "_1",
            "_2",
            "_3"
          ]
        }
      ],
      "bb3": [
        "state:",
        {
          "accessible": [],
          "owned": []
        }
      ]
    }
  ],
  "bb2": [
    [
      [
        "state:",
        {
          "accessible": [
            "_1",
            "_2",
            "_3"
          ],
          "owned": [
            "_1",
            "_2",
            "_3"
          ]
        },
        "statement: (*((_1.1: (&mut u32, u32)).0: &mut u32)) = move (_3.0: u32)"
      ],
      [
        "state:",
        {
          "accessible": [
            "_1",
            "_2",
            "_3"
          ],
          "owned": [
            "_1",
            "_2",
            "_3"
          ]
        },
        "statement: _0 = const ()"
      ]
    ],
    "state before terminator:",
    {
      "accessible": [
        "_0",
        "_1",
        "_2",
        "_3"
      ],
      "owned": [
        "_0",
        "_1",
        "_2",
        "_3"
      ]
    },
    "terminator: return",
    {}
  ],
  "bb3": [
    [],
    "state before terminator:",
    {
      "accessible": [],
      "owned": []
    },
    "terminator: resume",
    {}
  ]
}
